### PR TITLE
[CCSTAK-732] Adding conversion data to staking rewards

### PR DESCRIPTION
**Pull request by shreifabdallah**
### What changed? Why?

- Added `conversion_price` & `conversion_time` to the `StakingReward` task
- Added `conversion_price` & `conversion_time` to `staking_reward_test.ts`

### Testing notes

```
import { Coinbase, StakingReward } from "./src";
const coinbase = Coinbase.configureFromJson({ filePath: "your key path" });
let rewards = await StakingReward.list( Coinbase.networks.EthereumMainnet , Coinbase.assets.Eth,["0x60c7e246344ae3856cf9abe3a2e258d495fc39e0"],"2024-05-01T00:00:00Z","2024-05-21T00:00:00Z");
rewards.forEach(reward => {console.log(reward.toString())});


> rewards.forEach(reward => {console.log(reward.toString())});


StakingReward { date: '2024-05-01T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '3.61' usd_value: '3.61' conversion_price: '2971.419922' conversion_time: '2024-05-02T00:09:00.000Z' }
StakingReward { date: '2024-05-02T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.03' usd_value: '2.03' conversion_price: '2984.810059' conversion_time: '2024-05-03T00:09:00.000Z' }
StakingReward { date: '2024-05-03T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.26' usd_value: '2.26' conversion_price: '3103.570068' conversion_time: '2024-05-04T00:09:00.000Z' }
StakingReward { date: '2024-05-04T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.13' usd_value: '2.13' conversion_price: '3116.860107' conversion_time: '2024-05-05T00:09:00.000Z' }
StakingReward { date: '2024-05-05T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.92' usd_value: '1.92' conversion_price: '3144.01001' conversion_time: '2024-05-06T00:09:00.000Z' }
StakingReward { date: '2024-05-06T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.9' usd_value: '1.9' conversion_price: '3070.639893' conversion_time: '2024-05-07T00:09:00.000Z' }
StakingReward { date: '2024-05-07T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.83' usd_value: '1.83' conversion_price: '3016.580078' conversion_time: '2024-05-08T00:09:00.000Z' }
StakingReward { date: '2024-05-08T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.88' usd_value: '1.88' conversion_price: '2976.870117' conversion_time: '2024-05-09T00:09:00.000Z' }
StakingReward { date: '2024-05-09T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.15' usd_value: '2.15' conversion_price: '3033.02002' conversion_time: '2024-05-10T00:09:00.000Z' }
StakingReward { date: '2024-05-10T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.61' usd_value: '1.61' conversion_price: '2908.919922' conversion_time: '2024-05-11T00:09:00.000Z' }
StakingReward { date: '2024-05-11T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.88' usd_value: '1.88' conversion_price: '2914.639893' conversion_time: '2024-05-12T00:09:00.000Z' }
StakingReward { date: '2024-05-12T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.81' usd_value: '1.81' conversion_price: '2929.969971' conversion_time: '2024-05-13T00:09:00.000Z' }
StakingReward { date: '2024-05-13T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.91' usd_value: '1.91' conversion_price: '2944.370117' conversion_time: '2024-05-14T00:09:00.000Z' }
StakingReward { date: '2024-05-14T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.22' usd_value: '2.22' conversion_price: '2884.929932' conversion_time: '2024-05-15T00:09:00.000Z' }
StakingReward { date: '2024-05-15T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2.32' usd_value: '2.32' conversion_price: '3035' conversion_time: '2024-05-16T00:09:00.000Z' }
StakingReward { date: '2024-05-16T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '2' usd_value: '2' conversion_price: '2945.51001' conversion_time: '2024-05-17T00:09:00.000Z' }
StakingReward { date: '2024-05-17T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.85' usd_value: '1.85' conversion_price: '3092.560059' conversion_time: '2024-05-18T00:09:00.000Z' }
StakingReward { date: '2024-05-18T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.99' usd_value: '1.99' conversion_price: '3121.060059' conversion_time: '2024-05-19T00:09:00.000Z' }
StakingReward { date: '2024-05-19T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '1.78' usd_value: '1.78' conversion_price: '3074.280029' conversion_time: '2024-05-20T00:09:00.000Z' }
StakingReward { date: '2024-05-20T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '3.02' usd_value: '3.02' conversion_price: '3680.570068' conversion_time: '2024-05-21T00:09:00.000Z' }
```

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: src/client/api.ts [hunk 1]
```diff
@@ -757,6 +782,37 @@ export interface GetStakingContextRequest {
      */
     'options': { [key: string]: string; };
 }
+/**
+ * The balance of an asset onchain at a particular block
+ * @export
+ * @interface HistoricalBalance
+ */
+export interface HistoricalBalance {
```
Comment (ProfMoo) [inline src/client/api.ts:790]: Is this new object part of the PR? Or unanticipated generated changes that came along with the proposed changes? If this was intentional, then `HistoricalBalance` might be too vague for what we are adding here. Perhaps it'd be more accurate to call this `HistoricalStakingBalance`?

Comment (shreifabdallah) [reply src/client/api.ts:790]: > unanticipated generated changes that came along with the proposed changes

This is it.

File: src/coinbase/staking_reward.ts [hunk 2]
```diff
@@ -109,12 +109,39 @@ export class StakingReward {
...
+  }
+
   /**
    * Print the Staking Reward as a string.
    *
    * @returns The string representation of the Staking Reward.
    */
   public toString(): string {
-    return `StakingReward { date: '${this.date().toISOString()}' address: '${this.addressId()}' amount: '${this.amount().toString()}' }`;
+    return `StakingReward { date: '${this.date().toISOString()}' address: '${this.addressId()}' amount: '${this.amount().toString()}' usd_value: '${this.usdValue().toString()}' conversion_price: '${this.conversionPrice().toString()}' conversion_time: '${this.conversionTime().toISOString()}' }`;
```
Comment (ProfMoo) [inline src/coinbase/staking_reward.ts:145]: Does this function call still print out logically in the case where a user asks for native units?

Comment (shreifabdallah) [reply src/coinbase/staking_reward.ts:145]: I'll add a test case for this

But it'll always print out conversion data regardless of the user provided format.

Comment (shreifabdallah) [reply src/coinbase/staking_reward.ts:145]: Would you like it to behave differently if user asks for native units (not provide conversion data?)

Comment (shreifabdallah) [reply src/coinbase/staking_reward.ts:145]: This is how it looks when the user supplies `native`

```
> let rewards = await StakingReward.list( Coinbase.networks.EthereumMainnet , Coinbase.assets.Eth,["0x60c7e246344ae3856cf9abe3a2e258d495fc39e0"],"2024-05-01T00:00:00Z","2024-05-21T00:00:00Z", "native");
rewards.forEach(reward => {console.log(reward.toString())});undefined
> rewards.forEach(reward => {console.log(reward.toString())});
StakingReward { date: '2024-05-01T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.001212525541415161' usd_value: '3.61' conversion_price: '2971.419922' conversion_time: '2024-05-02T00:09:00.000Z' }
StakingReward { date: '2024-05-02T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000679464305235174' usd_value: '2.03' conversion_price: '2984.810059' conversion_time: '2024-05-03T00:09:00.000Z' }
StakingReward { date: '2024-05-03T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000726030823305604' usd_value: '2.26' conversion_price: '3103.570068' conversion_time: '2024-05-04T00:09:00.000Z' }
StakingReward { date: '2024-05-04T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000681550080574926' usd_value: '2.13' conversion_price: '3116.860107' conversion_time: '2024-05-05T00:09:00.000Z' }
StakingReward { date: '2024-05-05T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000609926334576295' usd_value: '1.92' conversion_price: '3144.01001' conversion_time: '2024-05-06T00:09:00.000Z' }
StakingReward { date: '2024-05-06T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000618244877949755' usd_value: '1.9' conversion_price: '3070.639893' conversion_time: '2024-05-07T00:09:00.000Z' }
StakingReward { date: '2024-05-07T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000604674331782671' usd_value: '1.83' conversion_price: '3016.580078' conversion_time: '2024-05-08T00:09:00.000Z' }
StakingReward { date: '2024-05-08T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000631405122359757' usd_value: '1.88' conversion_price: '2976.870117' conversion_time: '2024-05-09T00:09:00.000Z' }
StakingReward { date: '2024-05-09T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.00070716574809992' usd_value: '2.15' conversion_price: '3033.02002' conversion_time: '2024-05-10T00:09:00.000Z' }
StakingReward { date: '2024-05-10T00:00:00.000Z' address: '0x60c7e246344ae3856cf9abe3a2e258d495fc39e0' amount: '0.000552724206304093' usd_value: '1.61' conversion_price: '2908.919922' conversion_time: '2024-05-11T00:09:00.000Z' }
```

Comment (ProfMoo) [reply src/coinbase/staking_reward.ts:145]: I think we should avoid something like this:

```
date: <date> address: <address> count: <count> usd_value:  conversion_price:  conversion_time:
```

getting printed, where the USD information has empty values. As long as we're not printing our empty values, then I'm good to approve 👍.

Comment (shreifabdallah) [reply src/coinbase/staking_reward.ts:145]: Yeah it'll never be empty, RR always provides USD data, so does CDP service

**Review by ProfMoo — APPROVED** [review]
LGTM!

**Review by John-peterson-coinbase — APPROVED** [review]

**Pull request merged**